pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,476 @@
+		}
+
+		for _, uid := range found {
+			if err := m.recordPodMappings(uid); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return &m, nil
+}
+
+// recordPodMappings registers the range used for the user namespace iif the
+// usernsConfFile exits in the pod directory.
+func (m *usernsManager) recordPodMappings(pod types.UID) error {
+	usernsConfFile := m.getUserNamespaceMappingsFile(pod)
+
+	// If the size is too big, something is fishy. Don't read it all in mem.
+	// Ignore errors, the file might not be created yet and afterwards we
+	// check which error was returned.
+	if info, err := os.Stat(usernsConfFile); err == nil && info.Size() >= 1024*1024 {
+		return fmt.Errorf("pod user namespace file is too big: %q", usernsConfFile)
+	}
+
+	content, err := os.ReadFile(usernsConfFile)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if string(content) == "" {
+		return nil
+	}
+
+	_, err = m.parseUserNsFile(pod, content)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getUserNamespaceMappingsFile returns the path to the file that contains the user namespace configuration.
+func (m *usernsManager) getUserNamespaceMappingsFile(pod types.UID) string {
+	if m.kl == nil {
+		return ""
+	}
+	return path.Join(m.kl.getPodDir(pod), "userns")
+}
+
+// IsSet checks if the specified index is already set.
+func (m *usernsManager) isSet(v uint32) bool {
+	index := v / userNsLength
+	return m.used.isSet(index)
+}
+
+// allocateOne finds a free user namespace and allocate it to the specified pod.
+// The first return value is the first ID in the user namespace, the second returns
+// the length for the user namespace range.
+func (m *usernsManager) allocateOne(pod string) (uint32, uint32, error) {
+	firstZero, err := m.used.findAndSetFirstZero()
+	if err != nil {
+		return 0, 0, fmt.Errorf("could not allocate user namespace: %v", err)
+	}
+	firstID := firstZero * userNsLength
+	m.usedBy[pod] = firstID
+	return firstID, userNsLength, nil
+}
+
+// record stores the user namespace [from; from+length] to the specified pod.
+func (m *usernsManager) record(pod types.UID, from, length uint32) error {
+	if length != userNsLength {
+		return fmt.Errorf("wrong user namespace length %v", length)
+	}
+	if from%userNsLength != 0 {
+		return fmt.Errorf("wrong user namespace offset specified %v", from)
+	}
+	if prevFrom, found := m.usedBy[string(pod)]; found && prevFrom != from {
+		return fmt.Errorf("user namespace range already used by pod %q", pod)
+	}
+
+	// "from" is a ID (UID/GID), set the corresponding userns of size
+	// userNsLength in the bit-array.
+	m.used.set(from / userNsLength)
+	m.usedBy[string(pod)] = from
+	return nil
+}
+
+// Release releases the user namespace allocated to the specified pod.
+func (m *usernsManager) release(pod string) {
+	v, ok := m.usedBy[pod]
+	if !ok {
+		return
+	}
+	delete(m.usedBy, pod)
+
+	m.currPods--
+	m.removed++
+	if m.removed%1000 == 0 {
```

Comment:
I'd name the constant, rather than a literal `1000`.

Reply:
it is used to avoid memory leaks since memory used internally by go maps is never freed

I'll add a comment to explain that

Reply:
Fixed

Reply:
`const mapReInitializeThreshold = 1000`